File: internal/models/speech_usage.go
// Copyright (c) 2026 LingByte. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0

package models

import "time"

// OpenAPI 语音用量种类（与 speech_usage.kind 一致）。
const (
	SpeechUsageKindASR = "asr"
	SpeechUsageKindTTS = "tts"
)

// SpeechUsage 记录 OpenAPI 语音（ASR/TTS）单次调用，便于审计与统计；不落原始音频 base64。
type SpeechUsage struct {
	ID               string    `json:"id" gorm:"primaryKey;type:varchar(64)"`
	RequestID        string    `json:"request_id" gorm:"type:varchar(64);uniqueIndex;not null"`
	CredentialID     int       `json:"credential_id" gorm:"index"`
	UserID           string    `json:"user_id" gorm:"type:varchar(64);index"`
	Kind             string    `json:"kind" gorm:"type:varchar(16);not null;index"`
	Provider         string    `json:"provider" gorm:"type:varchar(64);index"`
	ChannelID        int       `json:"channel_id" gorm:"index"`
	Group            string    `json:"group" gorm:"type:varchar(128)"`
	RequestType      string    `json:"request_type" gorm:"type:varchar(64);not null;index"`
	RequestContent   string    `json:"request_content" gorm:"type:text"`
	ResponseContent  string    `json:"response_content" gorm:"type:text"`
	LatencyMs        int64     `json:"latency_ms"`
	StatusCode       int       `json:"status_code"`
	Success          bool      `json:"success"`
	ErrorMessage     string    `json:"error_message" gorm:"type:text"`
	AudioInputBytes  int64     `json:"audio_input_bytes"`
	AudioOutputBytes int64     `json:"audio_output_bytes"`
	TextInputChars   int       `json:"text_input_chars"`
	UserAgent        string    `json:"user_agent" gorm:"type:varchar(500)"`
	IPAddress        string    `json:"ip_address" gorm:"type:varchar(45)"`
	RequestedAt      time.Time `json:"requested_at" gorm:"not null;index"`
	CompletedAt      time.Time `json:"completed_at" gorm:"index"`
	CreatedAt        time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt        time.Time `json:"updated_at" gorm:"autoUpdateTime"`
}

func (SpeechUsage) TableName() string {
	return "speech_usage"
}

File: web/src/api/speechUsage.ts
import { assertOk, type Paginated } from '@/api/mailAdmin'
import { get } from '@/utils/request'

export interface SpeechUsageRow {
  id: string
  request_id: string
  credential_id: number
  user_id: string
  kind: string
  provider: string
  channel_id: number
  group: string
  request_type: string
  request_content: string
  response_content: string
  latency_ms: number
  status_code: number
  success: boolean
  error_message: string
  audio_input_bytes: number
  audio_output_bytes: number
  text_input_chars: number
  user_agent: string
  ip_address: string
  requested_at: string
  completed_at: string
  created_at: string
  updated_at: string
}

export type SpeechUsageListParams = {
  page?: number
  pageSize?: number
  user_id?: string
  kind?: string
  credential_id?: number
  channel_id?: number
  request_id?: string
  provider?: string
  request_type?: string
  success?: boolean
  from?: string
  to?: string
}

function toQuery(p: SpeechUsageListParams): string {
  const q = new URLSearchParams()
  if (p.page != null) q.set('page', String(p.page))
  if (p.pageSize != null) q.set('pageSize', String(p.pageSize))
  if (p.user_id) q.set('user_id', p.user_id)
  if (p.kind) q.set('kind', p.kind)
  if (p.credential_id != null && p.credential_id > 0) q.set('credential_id', String(p.credential_id))
  if (p.channel_id != null && p.channel_id > 0) q.set('channel_id', String(p.channel_id))
  if (p.request_id) q.set('request_id', p.request_id)
  if (p.provider) q.set('provider', p.provider)
  if (p.request_type) q.set('request_type', p.request_type)
  if (p.success === true) q.set('success', 'true')
  if (p.success === false) q.set('success', 'false')
  if (p.from) q.set('from', p.from)
  if (p.to) q.set('to', p.to)
  const s = q.toString()
  return s ? `?${s}` : ''
}

export async function listSpeechUsage(params: SpeechUsageListParams): Promise<Paginated<SpeechUsageRow>> {
  const r = await get<Paginated<SpeechUsageRow>>(`/api/speech-usage${toQuery(params)}`)
  return assertOk(r)
}

export async function getSpeechUsage(id: string): Promise<SpeechUsageRow> {
  const enc = encodeURIComponent(id)
  const r = await get<{ usage: SpeechUsageRow }>(`/api/speech-usage/${enc}`)
  const d = assertOk(r)
  return d.usage
}
